src/lib/FreeC/Pass/SharingAnalysisPass.hs

```diff
@@ -0,0 +1,144 @@
+
+import           Control.Monad             ( foldM, mplus )
+import           Data.Map.Strict           ( Map )
+import qualified Data.Map.Strict        as Map
+
+import           FreeC.Environment.Fresh   ( freshHaskellQName )
+import           FreeC.Environment.Renamer ( renameAndDefineVar )
+import qualified FreeC.IR.Syntax        as IR
+import           FreeC.IR.Subst
+import           FreeC.IR.SrcSpan
+import           FreeC.Monad.Converter
+import           FreeC.Pass
+
+-- | Checks all function declarations if they contain variables that occur
+--   multiple times on the same right-hand side.
+--   If that is the case, a @let@-expression is introduced that binds the
+--   variables to a fresh ones and replaces the occurrences with the newly
+--   introduced variable.
+sharingAnaylsisPass :: Pass IR.Module
+sharingAnaylsisPass
+    (IR.Module srcSpan name imports typeDecls typeSigs pragmas funcDecls) = do
+        funcDecls' <- mapM analyseSharingDecl (funcDecls)
+        return (IR.Module srcSpan name imports typeDecls typeSigs pragmas
+                funcDecls')
+
+-- | Checks a function declaration if it contains a variable that occurs
+--   multiple times on the right-hand side.
+--   If that is the case, a @let@-expression is introduced.
+analyseSharingDecl :: IR.FuncDecl -> Converter IR.FuncDecl
+analyseSharingDecl (IR.FuncDecl srcSpan ident typeArgs args returnType rhs) = do
+    let varList = (map (\p -> ( fst p, snd $ snd p )) . filter
+                   ((> 1) . fst . snd) . Map.toList . countVarNames) rhs
+    rhs' <- buildLet rhs varList
+    return (IR.FuncDecl srcSpan ident typeArgs args returnType rhs')
+
+-- | Builds a @let@-expression from the given expression and variable names.
+--
+--   Computes @let@-bindings from the given variables, composes the resulting
+--   substitutions and applies the substitution on the expression.
+buildLet
+    :: IR.Expr -> [ ( IR.VarName, Maybe IR.TypeScheme ) ] -> Converter IR.Expr
+buildLet expr []   = return expr
+buildLet expr vars = do
+    let srcSpan = IR.exprSrcSpan expr
+    ( binds, substs ) <- buildBinds srcSpan vars
+    return (IR.Let srcSpan binds (applySubst (composeSubsts substs) expr)
+            (IR.exprTypeScheme expr))
+
+-- | Converts the list containing variables into @let@-bindings where
+--   the variable pattern is a fresh variable and the right-hand side is
+--   a variable that occurred multiple times on right hand sides.
+--
+--   Also computes substitutions mapping given variables to fresh variables.
+--   Returns the generated let-bindings and the substitutions.
+buildBinds :: SrcSpan -> [ ( IR.VarName, Maybe IR.TypeScheme ) ]
+    -> Converter ( [ IR.Bind ], [ Subst IR.Expr ] )
+buildBinds srcSpan = foldM buildBind ( [], [] )
+  where
+    buildBind :: ( [ IR.Bind ], [ Subst IR.Expr ] )
+        -> ( IR.VarName, Maybe IR.TypeScheme )
+        -> Converter ( [ IR.Bind ], [ Subst IR.Expr ] )
```

Comment:
I think this fold is just [`mapAndUnzipM`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Monad.html#v:mapAndUnzipM) from `Control.Monad`.

```suggestion
buildBinds srcSpan = mapAndUnzipM buildBind
  where
    buildBind
        :: (IR.VarName, Maybe IR.TypeScheme)
        -> Converter (IR.Bind, Subst IR.Expr)
```